### ISSUE TITLE
enable custom Generator classes

### ISSUE THREAD

**Issue by YehudaCorsia** [wolverdude/GenSON#27]
I want add more functionality (like min/max in numbers and length in strings) , and I am search for ask you some questions.

Please add your mails or contact me soon. 

My mail is Yehudacorsia@gmail.com

**Comment by wolverdude** [comment]
Hi @YehudaCorsia, we can discuss things openly here. That's what issue threads are for. What specific features or issues would you like to be addressed?

**Comment by YehudaCorsia** [comment]
OK, I want add support in Min/Max in length of strings and also in numbers.

What are you think about this?

**Comment by wolverdude** [comment]
I think that adding that specific feature would be fairly simple. All you would need to do is update [this class](https://github.com/wolverdude/GenSON/blob/v1.0.2/genson/schema/generators/scalar.py#L45) so that `#add_schema`, `#add_object`, and `#to_schema deal with those keywords.

However, to make that change to the project itself would likely confuse other users because the project does not deal with keywords like `min` and `max` anywhere else and has not historically done so. For this reason, I won't approve that change unless it is made across the board with most if not all keywords which would be a big a large change in the project requiring a major version release.

I think there is a better solution that could solve your use-case without creating problems for other people. I could extend the API to enable custom `SchemaGenerator` classes. This way, you could create your own custom `SchemaBuilder` class with your own custom `SchemaGenerator` that handles your specific use-case. GenSON itself could even be extended on top of this to create a fully-featured keyword-handling `AllKeywordSchemaBuilder` class while maintain the current no-frills `SchemaBuilder` as the default.

**Comment by YehudaCorsia** [comment]
Thank you very much, I know it and I did my version which I create this functions that you talk about him,  in number Object and in string Object. If you want I can show you the code :-) 

I think that if you will allow the excellent idea about extend Builder Object to work with custom generators it's will be great. 

and i'll be happy to see this happens,  because also in my case I want to keep it as an option and not as something fixed.

And again thank you so much, this is very useful project!

**Comment by wolverdude** [comment]
Okay. I will bump up the priority of that feature and put it in the next minor version release.

**Comment by YehudaCorsia** [comment]
Great, goodluck!

**Comment by YehudaCorsia** [comment]
Do you have a time estimate how long it will take to you?
I am considering whether to try develop it by myself soon...

**Comment by wolverdude** [comment]
I'm planning to do it next weekend.

**Comment by wolverdude** [comment]
I'd love your feedback on how the API should be structured. Here's an example of the very initial API I've come up with. I wrote it for [testing purposes](https://github.com/wolverdude/GenSON/blob/ec3669e/test/test_custom.py#L5-L14), so it's simple.

```python
class MaxTenGenerator(Number):
    KEYWORDS = tuple(list(Number.KEYWORDS) + ['maximum'])

    def to_schema(self):
        schema = super(MaxTenGenerator, self).to_schema()
        schema['maximum'] = 10
        return schema

MaxTenSchemaBuilder = custom_schema_builder([MaxTenGenerator])
```

I don't love it because the created classes are scoped within the `create_schema_builder` function (e.g: `genson.schema.builder.custom_schema_builder.<locals>._CustomSchemaBuilder`).

Below are two alternate versions. Of all these, I prefer the last one. I'm looking for feedback and am open to any suggestions though.

#### Explicitly defining a new `SchemaNode`

```python
class MaxTenGenerator(Number):
    KEYWORDS = Number.add_keywords('maximum')

    def to_schema(self):
        schema = super(MaxTenGenerator, self).to_schema()
        schema['maximum'] = 10
        return schema

class MaxTenSchemaNode(SchemaNode):
    GENERATOR_CLASSES = SchemaNode.add_generator_classes(MaxTenSchemaNode)

class MaxTenSchemaBuilder(SchemaBuilder):
    NODE_CLASS = MaxTenSchemaNode
```

#### Without explicitly defining a new `SchemaNode`

```python
class MaxTenGenerator(Number):
    KEYWORDS = Number.add_keywords('maximum')

    def to_schema(self):
        schema = super(MaxTenGenerator, self).to_schema()
        schema['maximum'] = 10
        return schema

class MaxTenSchemaBuilder(SchemaBuilder):
    NODE_CLASS = SchemaNode.with_generator_classes(MaxTenSchemaNode)
```

**Comment by shtutzim** [comment]
> I don't love it because the created classes are scoped within the `create_schema_builder` function (e.g: `genson.schema.builder.custom_schema_builder.<locals>._CustomSchemaBuilder`).


you wrote `create_schema_builder` but you mean `custom_schema_builder()`

**Comment by YehudaCorsia** [comment]
Hey, i am also not sure what is the best way to do this.

but i am think that i am also prefer the last one.

I'd love to hear what happen with it, as well as if you need help writing it's i would be happy to help.

Good luck.

**Comment by wolverdude** [comment]
I'm not satisfied with any of them, actually. I'm currently researching metaprogramming patterns that might make it easier to work with.

**Comment by wolverdude** [comment]
We can use metaclasses to simplify the interface and completely abstract away `SchemaNode`:

```
class MaxTenNumber(Number):
    KEYWORDS = ('maximum',)

    def to_schema(self):
        schema = super(MaxTenNumber, self).to_schema()
        schema['maximum'] = 10
        return schema

class MaxTenSchemaBuilder(SchemaBuilder):
    SCHEMA_TYPES = (MaxTenNumber,)
```

**Comment by YehudaCorsia** [comment]
In the words of Tim Peters:

> Metaclasses are deeper magic than 99% of users should ever worry about. If you wonder whether you need them, you don't (the people who actually need them know with certainty that they need them, and don't need an explanation about why).
> ... unless you're like me, and you enjoy learning about obscure parts of Python.

so i prefer another ways, what are you think about using Decorators?

**Comment by YehudaCorsia** [comment]
Hi, I would be happy to know what are you planning to do about this  :-)

**Comment by wolverdude** [comment]
Late but not never, the solution is up on the develop branch. Please check out [the docs](https://github.com/wolverdude/GenSON/blob/develop/README.rst#customizing-schemabuilder) and give any feedback you have.

**Comment by wolverdude** [comment]
Fixed in `v1.2.0`
